src/sentry/static/sentry/app/components/smartSearchBar/index.tsx

```diff
@@ -1336,14 +1350,14 @@ const InputButton = styled(Button)`
 const StyledDropdownLink = styled(DropdownLink)`
   display: none;
 
-  @media (max-width: ${p => p.theme.breakpoints[2]}) {
+  @media (max-width: ${commonTheme.breakpoints[2]}) {
```

Comment:
We should keep using theme from context - we try to never import theme, regardless if the theme value never changes. Otherwise we have inconsistent usage of theme and it will only grow as people copy and paste code around.

We also have the overhead of when do I import theme and when do I not import theme, I think the goal should be, you never import theme.

Reply:
The goal of this PR is to make the decision simple: if you only need common values (e.g. breakpoints), then import commonTheme. Otherwise, use withTheme.

I have pragmatic concerns with your proposal: first, my understanding is there is a small overhead associated with withTheme, at least an additional React component. There's also too many files directly importing theme: 77 imports (after this PR), 180 uses. As we see with SmartSearchBar, many of these are non-trivial to convert.

In addition to the pragmatic issues with forcing use of context, I don't understand what semantic issues exist with importing commonTheme if you need breakpoints or some other common value.

Reply:
> The goal of this PR is to make the decision simple: if you only need common values (e.g. breakpoints), then import commonTheme. Otherwise, use withTheme.

Wouldn't having one decision be even simpler: always use `withTheme`

> my understanding is there is a small overhead associated with withTheme, at least an additional React component

This is true, although impact is probably negligible.

> There's also too many files directly importing theme: 77 imports (after this PR), 180 uses. As we see with SmartSearchBar, many of these are non-trivial to convert.

Agreed there's still lots directly importing theme we shouldn't try to do them all in one swoop. However, of the (180-77) cases that you did change, `SmartSearchBar` seemed to be the exception?

I think we should be trying to remove the decision tree of how to use theme. Say a new developer joins and sees a file where we're importing commonTheme and thinks "ok, this is how we do this", and then later ends up trying to use a theme alias, which doesn't exist. Now they have to dig into theme.tsx to see what is an alias and what is common.

Reply:
> Agreed there's still lots directly importing theme we shouldn't try to do them all in one swoop. However, of the (180-77) cases that you did change, `SmartSearchBar` seemed to be the exception?

The numbers I posted before were unclear. After this PR, there are 77 imports of theme/commonTheme. In those 77 files, there are a total of 180 references to the imported theme. This PR removed 23, so extrapolating from that, there will be about 3 to 4 more "complicated" cases.

I'm very sensitive to the complications for new devs; like I said, the whole motivation of commit 4 is to remove the theme-is-silently-lightTheme footgun for new and old devs alike, which I think is important enough to justify the added complication. However, I just think that the costs of forcing everything onto withTheme (code churn + slight runtime overhead) doesn't justify the benefits of new devs being marginally less confused, especially because instead of silently breaking, it would be obviously not working, at which point they would be able to ask for help.

I think another option could be additional documentation, either in code comments or at develop.sentry.dev. Do you think that would be an acceptable alternative?

Reply:
I 100% agree with the intention and motivation and I think you're right that removing aliases from the default export will ultimately help, but the root issue is that theme is being used outside of the Theme Provider, and I think our goal should be to train developers to always use theme inside of a provider.

> I just think that the costs of forcing everything onto withTheme (code churn + slight runtime overhead) doesn't justify the benefits of new devs being marginally less confused

I disagree with this, I doubt the added runtime overhead will ever be a performance bottleneck for us, and saving devs from being confused is very valuable.

Code comments won't scale and would become noisy as we would need the same comments everywhere (and that assumes they get read). Having documentation is always welcomed, but I don't think it's something that devs will reach for unless prompted. Having an all or nothing approach scales: we can probably add a lint rule to disallow importing theme, with an error message explaining why (or pointing to develop).

Anyway I think the root root root issue is that we don't yet have tooling to show dark mode snapshots :)

Reply:
> Code comments won't scale and would become noisy as we would need the same comments everywhere (and that assumes they get read). Having documentation is always welcomed, but I don't think it's something that devs will reach for unless prompted. Having an all or nothing approach scales: we can probably add a lint rule to disallow importing theme, with an error message explaining why (or pointing to develop).

Hm, I think I didn't communicate clearly here. What I meant was that in order to mitigate your concern about someone not knowing how to use the new theme.tsx, we could put a comment at the top of theme.tsx saying "Sentry theming is explained at https://develop.sentry.dev/theming/" or something.

> I disagree with this, I doubt the added runtime overhead will ever be a performance bottleneck for us, and saving devs from being confused is very valuable.

I'm just not convinced that "all theme elements need this special procedure to be accessed because some of them may change depending on light/dark" is sufficiently simpler than "common theme elements can be accessed in the normal JS way, color aliases need this special procedure to be accessed due to light/dark" to justify the difficulty, added code complexity, and small runtime overhead. If we're talking about hypothetical future devs, I think "why do I need to use withTheme instead of just importing it directly" is also a plausible question.

What it comes down to is: I think this PR is good enough for now and can be merged in its current state. I think that migrating the remaining instances is much more work for much less benefit. However, if someone else does that work, I'm not opposed to merging it in a separate PR.